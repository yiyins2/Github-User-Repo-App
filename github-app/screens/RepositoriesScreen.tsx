import * as React from 'react';
import { StyleSheet, FlatList, ScrollView } from 'react-native';
import {
  ActivityIndicator, Card, Paragraph, Button,
} from 'react-native-paper';
import { useState, useEffect } from 'react';
import { Route } from '@react-navigation/native';
import { BottomTabNavigationProp } from '@react-navigation/bottom-tabs';
import { BottomTabParamList, RepositoriesParamList } from '../types';
import { View } from '../components/Themed';
import Repository from '../github-model/GithubRepo';
import Repositories from '../github-model/GithubRepos';
import getRepositories from '../github-model/repoQuery';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 35,
  },
});

/**
 * @param param0 navigation function and route params
 * @returns the Repositories screen
 */
export default function RepositoriesScreen({ navigation, route } :
{ navigation : BottomTabNavigationProp<BottomTabParamList, 'Repositories'>,
  route : Route<'Repositories', RepositoriesParamList['RepositoriesScreen']> }) : JSX.Element {
  const [repositories, setRepositories] = useState<null | Repositories>(null);
  const [errorMsg, setErrorMsg] = useState<null | string>(null);

  /**
   * Initialize the Repositories data
   */
  useEffect(() => {
    getRepositories(route.params?.login).then(
      (response : Repositories) => { setRepositories(null); setRepositories(response); },
      (error : Error) => { setErrorMsg(error.message); },
    );
  }, [route.params?.login]);

  if (errorMsg !== null) {
    return <View style={styles.container}>{errorMsg}</View>;
  }

  /**
   * @param props the card item
   * @returns the view of the list of Repository cards
   */
  function renderRepository(props : { item : Repository }) : JSX.Element {
    const repository = props.item;
    return (
      <Card style={{ margin: 10 }}>
        <Card.Title title={repository.name} />
        <Card.Content>
          <Button
            style={{ width: 300 }}
            icon="github"
            mode="outlined"
            onPress={() => navigation.navigate('Profile', { params: { login: repository.owner }, screen: 'ProfileScreen' })}
          >
            {repository.owner}
          </Button>
          <Paragraph>{repository.description}</Paragraph>
        </Card.Content>
      </Card>
    );
  }

  // render all cards
  return (
    <ScrollView>
      {repositories === null ? <View style={styles.container}><ActivityIndicator animating size="large" /></View>
        : (
          <FlatList
            data={repositories.repositories}
            renderItem={renderRepository}
            keyExtractor={(item) => item.name}
          />
        )}
    </ScrollView>
  );
}
