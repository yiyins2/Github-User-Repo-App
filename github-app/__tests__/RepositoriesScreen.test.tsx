/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/require-await */
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import axios from 'axios';
import RepositoriesScreen from '../screens/RepositoriesScreen';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

test('renders normal response correctly', async () => {
  mockedAxios.post.mockResolvedValue({
    data: {
      data: {
        user: {
          repositories: {
            totalCount: 2,
            nodes: [
              {
                name: 'a',
                owner: {
                  login: 'b',
                },
                description: 'c',
              },
              {
                name: 'd',
                owner: {
                  login: 'e',
                },
                description: 'f',
              },
            ],
          },
        },
      },
    },
  });
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<RepositoriesScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});

test('renders undefined response with activity indicator', async () => {
  mockedAxios.post.mockResolvedValue(undefined);
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<RepositoriesScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});

test('renders undefined error response with error message', async () => {
  mockedAxios.post.mockRejectedValue(new Error('Some Error Occurred'));
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<RepositoriesScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});
