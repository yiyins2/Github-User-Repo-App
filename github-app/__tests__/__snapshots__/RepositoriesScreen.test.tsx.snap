// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders normal response correctly 1`] = `
<RCTScrollView>
  <View>
    <RCTScrollView
      data={
        Array [
          Repository {
            "description": "c",
            "name": "a",
            "owner": "b",
          },
          Repository {
            "description": "f",
            "name": "d",
            "owner": "e",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderRadius": 4,
                "elevation": 1,
                "margin": 10,
                "shadowColor": "#000000",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.24,
                "shadowRadius": 0.75,
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flexGrow": 1,
                  "flexShrink": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingLeft": 16,
                    },
                    Object {
                      "minHeight": 50,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "flexDirection": "column",
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "500",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 20,
                              "letterSpacing": 0.15,
                              "lineHeight": 30,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "minHeight": 30,
                              },
                              Object {
                                "marginBottom": 2,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    a
                  </Text>
                </View>
                <View />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "paddingHorizontal": 16,
                    },
                    Object {
                      "paddingBottom": 16,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "rgba(0, 0, 0, 0.29)",
                      "borderRadius": 4,
                      "borderStyle": "solid",
                      "borderWidth": 0.5,
                      "elevation": 0,
                      "minWidth": 64,
                      "width": 300,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": undefined,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 4,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <View
                        style={
                          Object {
                            "marginLeft": 12,
                            "marginRight": -4,
                          }
                        }
                      >
                        <Text
                          accessibilityElementsHidden={true}
                          allowFontScaling={false}
                          importantForAccessibility="no-hide-descendants"
                          pointerEvents="none"
                          selectable={false}
                          style={
                            Array [
                              Object {
                                "color": "#6200ee",
                                "fontSize": 16,
                              },
                              Array [
                                Object {
                                  "lineHeight": 16,
                                  "transform": Array [
                                    Object {
                                      "scaleX": 1,
                                    },
                                  ],
                                },
                                Object {
                                  "backgroundColor": "transparent",
                                },
                              ],
                              Object {
                                "fontFamily": "material-community",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              Object {},
                            ]
                          }
                        >
                          ó°Š¤
                        </Text>
                      </View>
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              undefined,
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#6200ee",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        b
                      </Text>
                    </View>
                  </View>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "textAlign": "left",
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontSize": 14,
                            "letterSpacing": 0.25,
                            "lineHeight": 20,
                            "marginVertical": 2,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                >
                  c
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderRadius": 4,
                "elevation": 1,
                "margin": 10,
                "shadowColor": "#000000",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.24,
                "shadowRadius": 0.75,
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flexGrow": 1,
                  "flexShrink": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingLeft": 16,
                    },
                    Object {
                      "minHeight": 50,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "flexDirection": "column",
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "500",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 20,
                              "letterSpacing": 0.15,
                              "lineHeight": 30,
                              "marginVertical": 2,
                            },
                            Array [
                              Object {
                                "minHeight": 30,
                              },
                              Object {
                                "marginBottom": 2,
                              },
                              undefined,
                            ],
                          ],
                        ],
                      ]
                    }
                  >
                    d
                  </Text>
                </View>
                <View />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "paddingHorizontal": 16,
                    },
                    Object {
                      "paddingBottom": 16,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "rgba(0, 0, 0, 0.29)",
                      "borderRadius": 4,
                      "borderStyle": "solid",
                      "borderWidth": 0.5,
                      "elevation": 0,
                      "minWidth": 64,
                      "width": 300,
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": undefined,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 4,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "center",
                          },
                          undefined,
                        ]
                      }
                    >
                      <View
                        style={
                          Object {
                            "marginLeft": 12,
                            "marginRight": -4,
                          }
                        }
                      >
                        <Text
                          accessibilityElementsHidden={true}
                          allowFontScaling={false}
                          importantForAccessibility="no-hide-descendants"
                          pointerEvents="none"
                          selectable={false}
                          style={
                            Array [
                              Object {
                                "color": "#6200ee",
                                "fontSize": 16,
                              },
                              Array [
                                Object {
                                  "lineHeight": 16,
                                  "transform": Array [
                                    Object {
                                      "scaleX": 1,
                                    },
                                  ],
                                },
                                Object {
                                  "backgroundColor": "transparent",
                                },
                              ],
                              Object {
                                "fontFamily": "material-community",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              Object {},
                            ]
                          }
                        >
                          ó°Š¤
                        </Text>
                      </View>
                      <Text
                        numberOfLines={1}
                        selectable={false}
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Array [
                              Object {
                                "letterSpacing": 1,
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              undefined,
                              Object {
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "#6200ee",
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              Object {
                                "fontFamily": "System",
                                "fontWeight": "500",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        e
                      </Text>
                    </View>
                  </View>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "textAlign": "left",
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontSize": 14,
                            "letterSpacing": 0.25,
                            "lineHeight": 20,
                            "marginVertical": 2,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                >
                  f
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;

exports[`renders undefined error response with error message 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 35,
      },
    ]
  }
>
  Some Error Occurred
</View>
`;

exports[`renders undefined response with activity indicator 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 35,
      },
    ]
  }
>
  Cannot read property 'data' of undefined
</View>
`;
